### ISSUE TITLE
Check for existing built packages as early as possible

### ISSUE THREAD

**Issue by Photon89** [Jguer/yay#2260]
### Is your feature request related to a problem? Please describe.

When yay tries to build a package which is already built (for example when updating a split package), yay first makes a md5sum check, then unpacks the sources, then runs prepare()  and then realises that a built package already exists.

### Describe the solution you'd like

Why not doing the existence check as the very first step, thus, if the package already exists, all the unnecessary steps won't be executed.

### Describe alternatives you've considered

None

### Additional context

None

**Comment by jdholtz** [comment]
I did try moving the check before and, while it worked for normal AUR packages, it failed for `*-git` packages. The reason I found is that makepkg automatically runs `pkgver()` after extracting sources and updates the `pkgver` in the PKGBUILD. Therefore, it is necessary that the sources are extracted before the built package  check is done to ensure it is checking for the correct version.

Possibly just the prepare() function could be run after verifying that the package is not built, but I doubt that would have a speed increase. The main time-consumer is the extracting sources step which cannot be skipped due to my reasoning above.

**Comment by Photon89** [comment]
Maybe it is possible to check for already existing built packages without running makepkg? Say, parse the value of pkgver  from the PKGBUILD, and see if a pkg.tar.zst file for this version is already there? It think, this would work for git packages as well, as they write the version into the PKGBUILD when being built.

**Comment by jdholtz** [comment]
For git packages, the version that exists in the PKGBUILD is not correct until makepkg actually runs pkgver(). Attempting to run pkgver() without makepkg could cause problems, such as not checking out the correct branch that should have done in the prepare().

**Comment by Photon89** [comment]
Yes, exactly, but as we are trying to get rid of unnecessary rebuilds, the package has been already built once, so the version in the PKGBUILD is actually correct, isn't it?

**Comment by jdholtz** [comment]
Good point. Unforutnately, Yay runs these commands before displaying the clean menu, so the repository information would be reset:
```shell
/usr/bin/git -C <path-to-package> reset --hard HEAD
/usr/bin/git -C <path-to-package> merge --no-edit --ff
```

I think this is necessary in order to check for any updates the package has.

**Comment by Photon89** [comment]
Wait, I'm not sure, I can follow. :smiley: Why is the repository information getting reset?

**Comment by jdholtz** [comment]
> Why is the repository information getting reset?

Not resetting would cause a conflict in Git when executing the merge command. In order to pull the latest updates to install the package, Yay needs to pull from the remote repository (which needs the git reset).

**Comment by Photon89** [comment]
I'm not very proficient with git (a fun thing to say on github :smile: ) but I tried to do some research and according to https://stackoverflow.com/questions/42691131/get-latest-commits-on-another-branch-git-without-checking-it-out you could do a `git fetch` which seems to require no reset. Maybe it can work this way?

**Comment by jdholtz** [comment]
That could work. I’m not too familiar with how Yay deals with git repositories internally. 

@Jguer would the above suggestion (git fetch) be possible in Yay to help improve installation time for already-built packages?

**Comment by Photon89** [comment]
Activity!

**Comment by Photon89** [comment]
Activity!

**Comment by Photon89** [comment]
Activity!
